client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java

```diff
@@ -0,0 +1,237 @@
+             MergeManager<K,V> merger,
+             Progress progress,
+             Reporter reporter, ShuffleClientMetrics metrics,
+             ExceptionReporter exceptionReporter,
+             ShuffleReadClient shuffleReadClient,
+             long totalBlockCount) {
+    this.jobConf = job;
+    this.reporter = reporter;
+    this.status = status;
+    this.merger = merger;
+    this.progress = progress;
+    this.metrics = metrics;
+    this.exceptionReporter = exceptionReporter;
+    this.reduceId = reduceId;
+    ioErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.IO_ERROR.toString());
+    wrongLengthErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.WRONG_LENGTH.toString());
+    badIdErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.BAD_ID.toString());
+    wrongMapErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.WRONG_MAP.toString());
+    connectionErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.CONNECTION.toString());
+    wrongReduceErrs = reporter.getCounter(SHUFFLE_ERR_GRP_NAME,
+      RssFetcher.ShuffleErrors.WRONG_REDUCE.toString());
+
+    this.shuffleReadClient = shuffleReadClient;
+    this.totalBlockCount = totalBlockCount;
+  }
+
+  public void fetchAllBlocks() throws IOException, InterruptedException {
+    while (!stopped) {
+      try {
+        // If merge is on, block
+        merger.waitForResource();
+        // Do shuffle
+        metrics.threadBusy();
+        copyFromRssServer();
+      } catch (Exception e) {
+        stopFetchAfterException(e);
+      } finally {
+        metrics.threadFree();
+      }
+    }
+  }
+
+  @VisibleForTesting
+  public void copyFromRssServer() throws IOException {
+    // fetch a block
+    final long startFetch = System.currentTimeMillis();
+    CompressedShuffleBlock compressedBlock = shuffleReadClient.readShuffleBlockData();
+    ByteBuffer compressedData = null;
+    if (compressedBlock != null) {
+      compressedData = compressedBlock.getByteBuffer();
+    }
+    long fetchDuration = System.currentTimeMillis() - startFetch;
+    readTime += fetchDuration;
+
+    // uncompress the block
+    if (compressedData != null) {
+      final long startDecompress = System.currentTimeMillis();
+      ByteBuffer uncompressedData = RssShuffleUtils.decompressData(
+        compressedData, compressedBlock.getUncompressLength());
+      unCompressionLength += compressedBlock.getUncompressLength();
+      long decompressDuration = System.currentTimeMillis() - startDecompress;
+      decompressTime += decompressDuration;
+
+      // Allocate a MapOutput (- either in-memory or on-disk) to put uncompressed block
+      final long startSerialization = System.currentTimeMillis();
+      TaskAttemptID mapId = MRRssUtils.createMRTaskAttemptId(
+        reduceId.getJobID(), TaskType.MAP, compressedBlock.getBlockId());
```

Comment:
wrong method use.

Reply:
Update to use a faked TaskAttemptId instead.